evm/src/modules/HostManager.sol

```diff
@@ -40,17 +45,19 @@ contract HostManager is IIsmpModule {
         _params.admin = address(0);
     }
 
-    function onAccept(PostRequest memory request) external onlyIsmpHost {
+    function onAccept(PostRequest calldata request) external onlyIsmpHost {
         // Only Hyperbridge can send requests to this module.
         require(request.source.equals(StateMachine.polkadot(_params.paraId)), "Unauthorized request");
 
-        // TODO: we should decode the payload based on the first byte in the request.body
-        if (false) {
+        // note, the below will revert with solidity error `Panic(0x21)`
+        OnAcceptActions action = OnAcceptActions(uint8(request.body[0]));
+        if (action == OnAcceptActions.Withdraw) {
             // This is where relayers can withdraw their fees.
-            WithdrawParams memory params = abi.decode(request.body, (WithdrawParams));
+            WithdrawParams memory params = abi.decode(request.body[1:], (WithdrawParams));
             IHostManager(_params.host).withdraw(params);
         } else {
-            HostParams memory params = abi.decode(request.body, (HostParams));
+            // i.e if (action == OnAcceptActions.SetHostParam)
```

Comment:
prefer if this isn't implicitly assumed and we indeed check the message type

Reply:
unresolved

Reply:
more gas?

Reply:
yes, more safety

Reply:
it's already safe, the 1 byte (uint8) identifier of action to take is casted into an enum type, if the number is > than the max variant id of the enum, it reverts, so else means OnAcceptActions.SetHostParam is the only possible outcome if action != to the first one